internal/repository/ent/coupon.go

```diff
@@ -0,0 +1,675 @@
+		"total_requested", len(ids),
+		"cached_count", len(cachedCoupons),
+		"uncached_count", len(uncachedIDs))
+
+	// If all coupons are cached, return them
+	if len(uncachedIDs) == 0 {
+		result := make([]*domainCoupon.Coupon, 0, len(ids))
+		for _, id := range ids {
+			if coupon, exists := cachedCoupons[id]; exists {
+				result = append(result, coupon)
+			}
+		}
+		SetSpanSuccess(span)
+		return result, nil
+	}
+
+	// Fetch uncached coupons from database
+	client := r.client.Querier(ctx)
+
+	entCoupons, err := client.Coupon.Query().
+		Where(
+			coupon.IDIn(uncachedIDs...),
+			coupon.TenantID(types.GetTenantID(ctx)),
+			coupon.EnvironmentID(types.GetEnvironmentID(ctx)),
+		).
+		All(ctx)
+
+	if err != nil {
+		SetSpanError(span, err)
+		return nil, ierr.WithError(err).
+			WithHint("Failed to batch get coupons").
+			WithReportableDetails(map[string]interface{}{
+				"coupon_ids": uncachedIDs,
+			}).
+			Mark(ierr.ErrDatabase)
+	}
+
+	// Convert to domain models and cache them
+	fetchedCoupons := make(map[string]*domainCoupon.Coupon)
+	for _, entCoupon := range entCoupons {
+		domainCoupon := domainCoupon.FromEnt(entCoupon)
+		fetchedCoupons[entCoupon.ID] = domainCoupon
+		r.SetCache(ctx, domainCoupon)
+	}
+
+	// Combine cached and fetched coupons in the original order
+	result := make([]*domainCoupon.Coupon, 0, len(ids))
+	for _, id := range ids {
+		if coupon, exists := cachedCoupons[id]; exists {
+			result = append(result, coupon)
+		} else if coupon, exists := fetchedCoupons[id]; exists {
+			result = append(result, coupon)
+		}
+		// Note: We don't add nil for missing coupons to maintain consistency
+		// Missing coupons are logged but not included in the result
+	}
+
+	r.log.Debugw("completed batch coupon fetch",
+		"requested_count", len(ids),
+		"fetched_count", len(entCoupons),
+		"result_count", len(result))
+
+	SetSpanSuccess(span)
+	return result, nil
+}
+
+func (r *couponRepository) Update(ctx context.Context, c *domainCoupon.Coupon) error {
+	client := r.client.Querier(ctx)
+
+	r.log.Debugw("updating coupon",
+		"coupon_id", c.ID,
+		"tenant_id", c.TenantID,
+		"name", c.Name,
+	)
+
+	// Start a span for this repository operation
+	span := StartRepositorySpan(ctx, "coupon", "update", map[string]interface{}{
+		"coupon_id": c.ID,
+		"name":      c.Name,
+	})
+	defer FinishSpan(span)
+
+	updateQuery := client.Coupon.Update().
+		Where(
+			coupon.ID(c.ID),
+			coupon.TenantID(c.TenantID),
+			coupon.EnvironmentID(types.GetEnvironmentID(ctx)),
+		).
+		SetName(c.Name).
+		SetType(string(c.Type)).
+		SetCadence(string(c.Cadence)).
+		SetUpdatedAt(time.Now().UTC()).
+		SetUpdatedBy(types.GetUserID(ctx))
+
+	// Handle optional fields
+	if c.Rules != nil {
+		updateQuery = updateQuery.SetRules(*c.Rules)
+	}
+	if c.AmountOff != nil {
+		updateQuery = updateQuery.SetAmountOff(*c.AmountOff)
+	}
+	if c.PercentageOff != nil {
+		updateQuery = updateQuery.SetPercentageOff(*c.PercentageOff)
+	}
+	if c.Metadata != nil {
```

Comment:
allow updating only the minimal / static fields
for now can keep just name and metadata